### PR TITLE
test: test_topology_ops: adapt to tablets

### LINKED ISSUE #17589: test: test_topology_ops: reenable background writes in tablets mode
#17585 reenables background writes in `test_topology_ops`, but only when the test is run without tablets. The test fails with tablets because of #17025. Once that issue if fixed, we should reenable background writes with tablets in `test_topology_ops`.

Ideally, before sending and merging the fix, we should investigate if `test_topology_ops` with tablets and background writes is flaky. Before #17585, without tablets and with background writes the test was flaky in CI (1-2 failures in 1000 runs in dev mode).

### PR DESCRIPTION
**Pull request by tchaikov**
in e7d4e080, we reenabled the background writes in this test, but when running with tablets enabled, background writes are still disabled because of #17025, which was fixed last week. so we can enable background writes with tablets.

in this change,

* background writes are enabled with tablets.
* increase the number of nodes by 1 so that we have enough nodes to fulfill the needs of tablets, which enforces that the number of replicas should always satisfy RF.
* pass rf to `start_writes()` explicitly, so we have less magic numbers in the test, and make the data dependencies more obvious.

Fixes #17589
Signed-off-by: Kefu Chai <kefu.chai@scylladb.com>

- [x] tablets related fix. no need to backport

### CODE REVIEW

**Comment by tchaikov** [comment]
the test passes locally. but it is failing on jenkins.. looking..

**Comment by kbr-scylla** [comment]
Restarted CI build

**Review by patjed41 — APPROVED** [review]
Looks good if the test passes in CI

**Comment by tchaikov** [comment]
the test is failing. looking..

**Comment by kbr-scylla** [comment]
But also for the case `tablets_enabled=false`, apparently.

But for `tablets_enabled=false` your change should be transparent, if I understand it correctly (please review and confirm that it is)

-- and if so, this would mean that the test was already flaky for `tablets_enabled=false` case :( (cc @patjed41)

**Comment by patjed41** [comment]
I see:
```
ERROR 2024-05-22 01:46:10,462 [shard 1:stmt] storage_proxy - No mapping for :: in the passed effective replication map
```
It's one of the "no ip mapping" issues or a new one. The similar issues are #18676 and #18843. This test fails with tablets disabled and no disrupted topology operations, so I guess it's a new one.

**Comment by kbr-scylla** [comment]
> It's one of the "no ip mapping" issues or a new one. The similar issues are https://github.com/scylladb/scylladb/issues/18676 and https://github.com/scylladb/scylladb/issues/18843. This test fails with tablets disabled and no disrupted topology operations, so I guess it's a new one.

Damn it :(
cc @gleb-cloudius

**Comment by gleb-cloudius** [comment]
> It could be the same root cause as #18843 -- that LEFT nodes are still inserted into `locator::topology` (but without IP mappings, which we don't have for LEFT nodes), IIUC this logic is generic for raft-topology (not only when tablets are enabled)
> 

No, without tablets the code is disabled. But I do not see any evidence yet that the failure here with and without tablets are the same. Logs without fail on timeout and with on something else. I will probably take 10 more years for me to download logs, so I cannot check better.

File: test/topology_experimental_raft/test_topology_ops.py
Comment (gleb-cloudius) [inline test/topology_experimental_raft/test_topology_ops.py:28]: Isn't ` tablets_enabled` is a string which is always true?

Comment (gleb-cloudius) [inline test/topology_experimental_raft/test_topology_ops.py:28]: Isn't `tablets_enabled` is a non empty string here which is always `True`?

**Comment by gleb-cloudius** [comment]
The test is in `topology_experimental_raft` suit which has tablets enabled by default since  6defcb7bd55c4ccd752297ad77df7a4200e42b37 (forever). So all tests are running with tablets enabled.

File: test/topology_experimental_raft/test_topology_ops.py
Comment (patjed41) [reply test/topology_experimental_raft/test_topology_ops.py:28]: Yes, and we have multiple (5 if I'm not wrong) tests broken in the same way. The first bug served as a bad example that caused the other bugs. We need to fix these tests and check if they pass.

The bigger problem with this particular test is that before this PR we didn't send writes if tablets were enabled. But they were always enabled, so we never sent writes. The test was basically disabled. This is very bad for such a basic and important test.

Hopefully, the situation is not that bad. IIRC, I verified that this test was not flaky before incorrectly parameterizing it with tablets around 3 month ago. So, I hope that "missing IP mapping" is the only regression.

@kbr-scylla

Comment (patjed41) [reply test/topology_experimental_raft/test_topology_ops.py:28]: > Yes, and we have multiple (5 if I'm not wrong) tests broken in the same way. The first bug served as a bad example that caused the other bugs. We need to fix these tests and check if they pass.

Only one test was broken -- `test_default_tombstone_gc`. I opened #18888.

I'll be investigating failures in this test after re-enabling writes. This PR will have to wait until all failures are fixed.

Comment (gleb-cloudius) [reply test/topology_experimental_raft/test_topology_ops.py:28]: The failures are because the test is always running with tablets (https://github.com/scylladb/scylladb/pull/18707#issuecomment-2132232745) and it is a known problem with tablets https://github.com/scylladb/scylladb/issues/18843

**Comment by tchaikov** [comment]
> @tchaikov you aborted twice: once to cancel test run, second time in the middle of artifacts collection phase, so we're missing a lot of logs unfortunately

if i don't abort the test the test node would be full of log file and running out of space. i spotted this the other time.

File: test/topology_experimental_raft/test_topology_ops.py
Comment (tchaikov) [reply test/topology_experimental_raft/test_topology_ops.py:28]: but my observation is quite the opposite , it fails with write load enabled and with tablets disabled. see #18932

**Comment by patjed41** [comment]
> > @tchaikov you aborted twice: once to cancel test run, second time in the middle of artifacts collection phase, so we're missing a lot of logs unfortunately
> 
> if i didn't abort the test, the test node would be full of log file and running out of space. i spotted this the other time.

#18851 fixed the huge log file issue.

Anyway, we now know that the test is flaky only with tablets enabled because of #18843. This PR is blocked by that issue.

File: test/topology_experimental_raft/test_topology_ops.py
Comment (gleb-cloudius) [reply test/topology_experimental_raft/test_topology_ops.py:28]: You need to move the test from topology_experimental_raft since this suit enables tablets by default. See https://github.com/scylladb/scylladb/pull/18707#issuecomment-2132232745.

Comment (patjed41) [reply test/topology_experimental_raft/test_topology_ops.py:28]: I fixed it in #18900 by changing the config.

**Comment by patjed41** [comment]
@tchaikov  #18900 should solve all problems in this test. So, if I'm not wrong, this PR is ready and is only waiting for a fix for #18843. Then, we can run CI again and see if it passes. If it does, we can merge this PR.

You are probably confused because I touched this test even though you are assigned to this issue. But because of the blunders fixed by #18900, we were running this test only with tablets and with write works disabled for 3 month. We wanted to verify raft-topology ASAP.

**Comment by tchaikov** [comment]
@patjed41 so, may i assign this ticket back to you? since you've done the heavy lifting.

**Comment by patjed41** [comment]
This issue is about ensuring that this test passes with tablets and write workers enabled. If the test fails again after fixing #18843, it should be investigated by someone with tablets knowledge (I don't really have it). Maybe we shouldn't change the assignment.

**Comment by tchaikov** [comment]
still pending on #18843

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Pull request closed without merging**
